flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -178,6 +179,8 @@
 	/** The kvState registration service in the task manager. */
 	private final KvStateService kvStateService;
 
+	private final List<Task> submittedTasks;
```

Comment:
At this point,  I would introduce a new class `TaskCompletionTracker` which keeps `submittedTasks` and has 2 methods: `trackTaskCompletion(Task task)` called from `submitTask` and `failIncompleteTasksAndWaitForAllCompleted()` which would return `FutureUtils.waitForAll(failSubmittedTasks())`.